File: 卡码/代码随想录/1/06-二叉树/Python/03-二叉树修改与构造/LeetCode_0106.py
#106. 从中序与后序遍历序列构造二叉树
from typing import *
from ..TreeNode import *

class Solution:
  def traversal(self,inOrder:List[int],inLeft:int,inRight:int,postOrder:List[int],postLeft:int,postRight:int)->Optional[TreeNode]:
    if inLeft>inRight:
      return None
    root:TreeNode=TreeNode(postOrder[postRight],None,None)
    if inLeft==inRight:
      return root
    rootIndex:int=0
    for i in range(inLeft,inRight+1):
      if inOrder[i]==root.val:
        rootIndex=i
        break
    leftLength:int=rootIndex-inLeft
    root.left=self.traversal(inOrder,inLeft,rootIndex-1,postOrder,postLeft,postLeft+leftLength-1)
    root.right=self.traversal(inOrder,rootIndex+1,inRight,postOrder,postLeft+leftLength,postRight-1)
    return root

  def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    if len(inorder)==0 or len(postorder)==0:
      return None
    return self.traversal(inorder,0,len(inorder)-1,postorder,0,len(postorder)-1)

File: 卡码/代码随想录/1/06-二叉树/C++/03-二叉树修改与构造/LeetCode_0106.cpp
//106. 从中序与后序遍历序列构造二叉树
#include <iostream>
#include <vector>
#include "../TreeNode.h"
using namespace std;

class Solution {
public:
	TreeNode* traversal(vector<int>& inOrder,int inLeft,int inRight,vector<int>& postOrder,int postLeft,int postRight){
		if(inLeft>inRight) return nullptr;
		TreeNode* root=new TreeNode(postOrder[postRight]);
		if(inLeft==inRight) return root;
		int rootIndex=0;
		for(int i=inLeft;i<=inRight;i++){
			if(inOrder[i]==root->val){
				rootIndex=i;
				break;
			}
		}
		int leftLength=rootIndex-inLeft;
		root->left=traversal(inOrder,inLeft,rootIndex-1,postOrder,postLeft,postLeft+leftLength-1);
		root->right=traversal(inOrder,rootIndex+1,inRight,postOrder,postLeft+leftLength,postRight-1);
		return root;
	}
  TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
		if(inorder.size()==0||postorder.size()==0) return nullptr;
		return traversal(inorder,0,inorder.size()-1,postorder,0,postorder.size()-1);
  }
};

int main(){
  system("pause");
  return 0;
}
